src/app/code/community/Shopgate/Cloudapi/Model/Api2/Customers/Addresses/Rest.php

```diff
@@ -0,0 +1,361 @@
+
+        try {
+            $address->save();
+        } catch (Mage_Core_Exception $e) {
+            $this->_error($e->getMessage(), Mage_Api2_Model_Server::HTTP_INTERNAL_ERROR);
+        } catch (Exception $e) {
+            $this->_critical(self::RESOURCE_INTERNAL_ERROR);
+        }
+
+        return $this->_getLocation($address);
+    }
+
+    /**
+     * Retrieve information about specified customer address
+     *
+     * @throws Mage_Api2_Exception
+     * @throws Exception
+     * @return array
+     */
+    protected function _retrieve()
+    {
+        /* @var $address Mage_Customer_Model_Address */
+        $address               = $this->_loadCustomerAddressById($this->getRequest()->getParam('id'));
+        $addressData           = $address->getData();
+        $addressData['street'] = $address->getStreet();
+
+        return $addressData;
+    }
+
+    /**
+     * Get customer addresses list
+     *
+     * @return array
+     * @throws Mage_Api2_Exception
+     */
+    protected function _retrieveCollection()
+    {
+        $data = array();
+        /* @var $address Mage_Customer_Model_Address */
+        foreach ($this->_getCollectionForRetrieve() as $address) {
+            $addressData                     = $address->getData();
+            $addressData['street']           = $address->getStreet();
+            $addressData['customAttributes'] = array();
+            foreach ($addressData as $code => $value) {
+                if (!in_array($code, $this->getDefaultAttributeList(), true)) {
+                    $addressData['customAttributes'][$code] = $value;
+                    unset($addressData[$code]);
+                }
+            }
+            $data[] = array_merge($addressData, $this->_getDefaultAddressesInfo($address));
+        }
+
+        return $data;
+    }
+
+    /**
+     * Retrieve collection instances
+     *
+     * @return Mage_Customer_Model_Resource_Address_Collection
+     * @throws Mage_Api2_Exception
+     * @throws Exception
+     */
+    protected function _getCollectionForRetrieve()
+    {
+        /* @var $customer Mage_Customer_Model_Customer */
+        $customer = $this->_loadCustomerById($this->getRequest()->getParam('customer_id'));
+
+        /* @var $collection Mage_Customer_Model_Resource_Address_Collection */
+        $collection = $customer->getAddressesCollection();
+
+        $this->_applyCollectionModifiers($collection);
+
+        return $collection;
+    }
+
+    /**
+     * Get array with default addresses information if possible
+     *
+     * @param Mage_Customer_Model_Address $address
+     *
+     * @return array
+     */
+    protected function _getDefaultAddressesInfo(Mage_Customer_Model_Address $address)
+    {
+        return array(
+            'is_default_billing'  => (int) $this->_isDefaultBillingAddress($address),
+            'is_default_shipping' => (int) $this->_isDefaultShippingAddress($address)
+        );
+    }
+
+    /**
+     * Update specified stock item
```

Comment:
I guess it's not updating stock item. :)

Reply:
lol